### PR TITLE
TeX on Markup widgets Div, Paragraph

### PR DESCRIPTION
**Pull request by IuryPiva**
```python
from bokeh.io import show
from bokeh.models import Div

div = Div(text="""This text includes some math $$\sin(x)$$ and is <b>html</b>!""", enable_tex=True, width=200, height=100)

show(div)
```
image

```ts
const div = new Div({
  text: `When \\(a \\ne 0\\), there are two solutions to \\(ax^2 + bx + c = 0\\) and they are
    $$x = {-b \\pm \\sqrt{b^2-4ac} \\over 2a}.$$`,
})

await display(div, [320, 120])
```
image

- [x] issues: relates to #11191 and #11355 
- [x] tests added / passed
- [x] release document entry (if new feature or API change)

Discussion on bundler and finding mathstrings #11589

### CODE REVIEW

**Comment by tcmetzger** [comment]
(Narrative) documentation will be included in https://github.com/bokeh/bokeh/pull/11544

File: bokeh/embed/bundle.py
Comment (mattpap) [inline bokeh/embed/bundle.py:428]: Instead of collecting all properties and their values, you should ask if a property exists and if so, just access it as a regular attribute, e.g.:
```py
return "enable_tex" in model.properties() and model.enable_tex
```

File: bokehjs/src/lib/models/text/mathjax/index.ts
Comment (mattpap) [inline bokehjs/src/lib/models/text/mathjax/index.ts:29]: This should have kept as "defaults" and later on you should use `{...defaults, ...options}`, which states the intent more clearly than this (when seeing this code I thought you would be adding more options, instead of just overriding the defaults).

File: bokehjs/src/lib/models/text/utils.ts
Comment (mattpap) [inline bokehjs/src/lib/models/text/utils.ts:8]: Snake case.

File: bokehjs/src/lib/models/text/utils.ts
Comment (mattpap) [inline bokehjs/src/lib/models/text/utils.ts:56]: ^^^

File: bokehjs/src/lib/models/text/utils.ts
Comment (mattpap) [inline bokehjs/src/lib/models/text/utils.ts:60]: What's the purpose of the first filter?

File: bokehjs/src/lib/models/text/utils.ts
Comment (mattpap) [inline bokehjs/src/lib/models/text/utils.ts:53]: You may want to look into `core/util/{array,arrayable}` and find some useful and more idiomatic functions to use instead of this (e.g. `sort_by()`).

File: bokehjs/src/lib/models/text/utils.ts
Comment (mattpap) [inline bokehjs/src/lib/models/text/utils.ts:3]: This needs to be extensively unit tested.

File: bokehjs/src/lib/models/text/mathjax/index.ts
Comment (mattpap) [inline bokehjs/src/lib/models/text/mathjax/index.ts:15]: This type (as well as `TexMacros`) should not be necessary to be duplicated here, as long as `mathjax.d.ts` is in scope. You just may need to use a namespace prefix to access those types. Also, is this TeX specific? If so, the name of this type should reflect this.

File: bokehjs/src/lib/models/widgets/div.ts
Comment (mattpap) [inline bokehjs/src/lib/models/widgets/div.ts:7]: The default provider must be shared between all models that use MathJax. Perhaps you may want to move `default_provider` to providers.

File: bokehjs/src/lib/models/widgets/div.ts
Comment (mattpap) [inline bokehjs/src/lib/models/widgets/div.ts:5]: You've just made `models/widgets/div` depend on **all** other model modules. Use a specific import.

File: bokehjs/src/lib/models/text/utils.ts
Comment (IuryPiva) [reply bokehjs/src/lib/models/text/utils.ts:60]: This will remove falsy values like empty strings, it was needed in a previous version of this method but not right now. i'm removing it.

File: bokeh/embed/bundle.py
Comment (bryevdv) [inline bokeh/embed/bundle.py:427]: This will also need to be impoved in the future, it can go along with the same other issue I mentioned creating in the other PR. For the immediate term I do think this would be preferred than mucking with string property names:
```
return isinstance(model, Markup) and model.enable_tex
```

File: bokeh/models/widgets/markups.py
Comment (bryevdv) [inline bokeh/models/widgets/markups.py:99]: more consistent with previous practice:
```suggestion
    Content can be interpreted as `TeX and LaTeX input`_ by  setting to True`
    
    .. _`TeX and LaTeX input`: https://docs.mathjax.org/en/latest/basic/mathematics.html#tex-and-latex-input
    
```

File: bokeh/models/widgets/markups.py
Comment (bryevdv) [inline bokeh/models/widgets/markups.py:111]: line length, also no need to code-ify True and False
```suggestion
    The default value is False, and only applies when rendering as HTML 
    (that is, when ``render_as_text`` is False)
```

File: bokeh/models/text.py
Comment (bryevdv) [inline bokeh/models/text.py:113]:
```suggestion
    Whether the math text is in display-mode or not (for TeX input). Default is False.
```

**Comment by tcmetzger** [comment]
As discussed with @bryevdv and @IuryPiva, the UX for LaTeX on Div widgets will be updated as follows:

- If the Div's `render_html` is true and there are MathJax default delimiter pairs in the text, then include the bundle and render the included LaTeX
- if the Div's `disable_math` is set to True, math processing is deactivated. This way, users can turn math processing off for a given div

This way, the MathJAx bundle will only be included if `disable_math` is `False` (default) and there are delimiter pairs detected (i.e. don't unconditionally include MathJax)

**Comment by bryevdv** [comment]
Just a slight refinement:

> This way, the MathJAx bundle will only be included if disable_math is False (default) and there are delimiter pairs detected *in at least one Div* (i.e. don't unconditionally include MathJax)

**Comment by bryevdv** [comment]
(sorry hit the wrong button, really wish there was a confirmation for that)

File: bokehjs/src/lib/models/text/math_text.ts
Comment (bryevdv) [inline bokehjs/src/lib/models/text/math_text.ts:415]: I actually can't find ref docs for `MathJax.tex2svg` despite searching. What is the undefined for (can you give a pointer to the docs)

File: bokehjs/src/lib/models/text/utils.ts
Comment (bryevdv) [inline bokehjs/src/lib/models/text/utils.ts:13]: Same comment as above, this is incredibly complicated and opaque manual text processing code. I don't recommend regex often but I don't see how this isn't a case to use non-greedy expressions for find all the pairs, more efficiently and with less code to maintain.

**Review by bryevdv — CHANGES_REQUESTED** [review]

**Comment by bryevdv** [comment]
@mattpap I am happy with the user-facing parts of this PR and the output generated. I think the implementation needs work but can be iterated on. To the extent you are able, please make it your highest priority to outline the *minimal* set of changes you require in the immediate term. This is the last thing blocking release.

File: bokeh/embed/util.py
Comment (IuryPiva) [reply bokeh/embed/util.py:370]: It is considered on BokehJS `find_math_parts` function, I do like your regex approach. I made this because I wanted something similar to what `find_math_parts` that can't be done with regexs

File: bokehjs/src/lib/models/text/utils.ts
Comment (IuryPiva) [reply bokehjs/src/lib/models/text/utils.ts:13]: So this is a shallow copy of what MathJax does. It is not a simple function to look for pair delimiters, it interprets them in order and save their positions on the text input. Maybe should have a better name.

File: bokehjs/src/lib/models/text/math_text.ts
Comment (IuryPiva) [reply bokehjs/src/lib/models/text/math_text.ts:415]: This should not be here. This second options are the [conversion options](http://docs.mathjax.org/en/latest/web/typeset.html?highlight=display%3A#conversion-options) documented on MathJax/index.ts too

File: bokeh/models/widgets/markups.py
Comment (mattpap) [inline bokeh/models/widgets/markups.py:83]: If I understand this correctly, this default will make bokeh always include mathjax if any markup is present? If that's the case, it's the other way that it should be.

File: bokehjs/src/lib/models/text/utils.ts
Comment (mattpap) [inline bokehjs/src/lib/models/text/utils.ts:10]: I think I already commented on this, but if not then use snake case.

File: bokehjs/src/lib/models/widgets/pretext.ts
Comment (mattpap) [inline bokehjs/src/lib/models/widgets/pretext.ts:13]: What's the point of those changes?

File: bokeh/embed/util.py
Comment (mattpap) [reply bokeh/embed/util.py:345]: Isn't the original styling what black would expect?

File: bokeh/models/widgets/markups.py
Comment (IuryPiva) [reply bokeh/models/widgets/markups.py:83]: You say it should be a default not processed for TeX input?

We discussed this UX before 
https://github.com/bokeh/bokeh/pull/11585#issuecomment-917338525

Comment (bryevdv) [reply bokeh/models/widgets/markups.py:83]: > If I understand this correctly, this default will make bokeh always include mathjax if any markup is present

yes, the expectation from users will be that it "just works" so this is the correct default behavior.

File: bokeh/embed/util.py
Comment (bryevdv) [reply bokeh/embed/util.py]: Well, we don't use Black. My suggestions bring these docstrings' formatting in line with every all the other docstrings already in the project. That said, I have made a decision this morning to no longer expend effort on formatting nits that are not covered by automation in some way (a test, or formatter). So, I leave it to your discretion whether to apply these.

File: bokehjs/src/lib/models/text/utils.ts
Comment (bryevdv) [reply bokehjs/src/lib/models/text/utils.ts:13]: can you link to the mathjax code? If this is truly what they do then I will withdraw my objection, but then we should leave comments with that link to the reference code we are drawing from.

Comment (IuryPiva) [reply bokehjs/src/lib/models/text/utils.ts:13]: https://github.com/mathjax/MathJax-src/blob/master/ts/input/tex/FindTeX.ts#L197

Comment (bryevdv) [reply bokehjs/src/lib/models/text/utils.ts:13]: I'm struggling a bit to see the direct connection. The MathJax code does use regexes (several calls to [`RegExp.exec`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/exec) throughout) but does not use `indexOf`, `map`, `filter`, or `sort` at all for searching. I guess we can revisit later as long as there is test coverage.

 But speaking of that, this function would be better if was decoupled from Bokeh models. It should *find* math parts, not find parts and also *convert* math parts. This function should return something like MathJax's `ProtoItem`, i.e. a thing that says *"Here's the text I found, here is what kind of text it is, here are the indices into the original text where I found it"* Then callers of `find_math_parts` can use that to convert as appropriate. That will be more straightforward to test (since checks can be by value, not by types).

**Comment by bryevdv** [comment]
@mattpap our stated plan was to release tomorrow. I have left my comments, and have communicated my minimal changes I want to see in `find_math_parts`. I am stating my intention to merge this PR tomorrow at noon (PST), unless there a specific Changes Requested in a review before then.

File: bokehjs/src/lib/models/widgets/markup.ts
Comment (mattpap) [inline bokehjs/src/lib/models/widgets/markup.ts:82]: `parts?.length > 0`

File: bokeh/models/widgets/markups.py
Comment (mattpap) [reply bokeh/models/widgets/markups.py:83]: It makes sense when coupled with `if contains_tex_string(model.text):`, which I didn't realise is the case. In this case it's fine. The point is that mathjax bundle shouldn't be loaded unnecessarily (given its size; by that I no only mean the download size, but also, and perhaps even more importantly, JS parsing/load time, which is significant).

File: bokehjs/src/lib/models/widgets/paragraph.ts
Comment (mattpap) [inline bokehjs/src/lib/models/widgets/paragraph.ts:18]: Try keeping `const` whenever possible (and avoid duplication at the same time), e.g.:
```ts
    const content = paragraph({style: {margin: 0}})

    if (this.has_math_disabled())
      content.textContent = this.model.text
    else
      content.innerHTML = this.process_tex()
```

File: bokehjs/test/integration/widgets.ts
Comment (mattpap) [inline bokehjs/test/integration/widgets.ts:219]: There should be an empty line between tests.

File: bokehjs/test/framework.ts
Comment (mattpap) [inline bokehjs/test/framework.ts:261]: Why this change? This error should be printed directly below a nicely formatted test description. If that not the case, then there's an async issue somewhere in the code.

File: bokehjs/src/lib/models/text/utils.ts
Comment (mattpap) [inline bokehjs/src/lib/models/text/utils.ts:12]: What's the reason for this being a case insensitive regular expression?

File: bokeh/models/widgets/markups.py
Comment (mattpap) [reply bokeh/models/widgets/markups.py:83]: Perhaps this property could be called `verbatim`, as in do not apply any additional processing to the contents of the `div`.

File: bokeh/models/widgets/markups.py
Comment (mattpap) [inline bokeh/models/widgets/markups.py:85]: We know what's the default value from the properties' system. There is no need for duplication, unless you want to explain more about the default case.

File: bokehjs/src/lib/models/widgets/markup.ts
Comment (mattpap) [inline bokehjs/src/lib/models/widgets/markup.ts:57]: String equality is not needed (and not encouraged in this codebase). Also please split off `return` into a new line.

File: bokehjs/src/lib/models/widgets/markup.ts
Comment (mattpap) [inline bokehjs/src/lib/models/widgets/markup.ts:24]: You still need to implement the rest of the protocol, the same as in `MathText`. This `await` doesn't guarantee that, e.g. a CDN bundle is loaded at this point (regardless whether we actually support CDN or not).

File: bokehjs/src/lib/models/widgets/markup.ts
Comment (mattpap) [inline bokehjs/src/lib/models/widgets/markup.ts:68]: Is this default value really needed? It seems to me that this (and some other instances of `.reduce()` at al) could be rewritten more clearly with a simple for loop.

File: bokeh/models/widgets/markups.py
Comment (bryevdv) [reply bokeh/models/widgets/markups.py:83]: I'm fine with a different name tho I wonder if users might think `verbatim` relates to `pre`

Comment (IuryPiva) [reply bokeh/models/widgets/markups.py:83]: I think `verbatim` would be a good replacement for `render_as_text`, since `disable_math` can still have `html` markups processed.

**Comment by mattpap** [comment]
We've just discussed the remaining required changes, so this should be finished soon.

**Comment by bryevdv** [comment]
Thanks @mattpap @IuryPiva for getting this over the finish line

**Comment by bryevdv** [comment]
@mattpap is this good to merge now? I would like to cut the RC today if possible

**Pull request merged**
